File: src/lib/chartLiveDaily.ts
/**
 * Append/replace today's daily bar from intraday session or Upstox LTP.
 */

import type { OHLCVCandle } from "@/hooks/useChartData";
import { aggregateSessionCandle } from "@/lib/fnoStrategies";
import { istDayKey } from "@/lib/taCompute";

export function todayIstDayKey(nowMs = Date.now()): string {
  return istDayKey(Math.floor(nowMs / 1000));
}

/** Unix seconds for 09:15 IST on a given YYYY-MM-DD day key. */
export function istSessionOpenUnix(dayKey: string): number {
  return Math.floor(new Date(`${dayKey}T09:15:00+05:30`).getTime() / 1000);
}

export function sessionCandlesForIstDay(candles: OHLCVCandle[], dayKey: string): OHLCVCandle[] {
  return candles.filter((c) => istDayKey(c.time) === dayKey);
}

export function dailySeriesMissingToday(daily: OHLCVCandle[]): boolean {
  if (!daily.length) return true;
  return istDayKey(daily[daily.length - 1].time) !== todayIstDayKey();
}

export function mergeIntradayForDay(
  daily: OHLCVCandle[],
  intraday: OHLCVCandle[],
  dayKey: string,
): OHLCVCandle[] {
  const session = sessionCandlesForIstDay(intraday, dayKey);
  const today = aggregateSessionCandle(session);
  if (!today) return daily;

  const existing = daily.find((c) => istDayKey(c.time) === dayKey);
  const time = existing?.time ?? istSessionOpenUnix(dayKey);
  const bar: OHLCVCandle = { ...today, time };
  const history = daily.filter((c) => istDayKey(c.time) !== dayKey);
  return [...history, bar].sort((a, b) => a.time - b.time);
}

export function appendTodayFromLtp(daily: OHLCVCandle[], ltp: number, dayKey?: string): OHLCVCandle[] {
  const key = dayKey ?? todayIstDayKey();
  if (daily.some((c) => istDayKey(c.time) === key)) {
    return mergeIntradayForDay(daily, [
      {
        time: istSessionOpenUnix(key),
        open: ltp,
        high: ltp,
        low: ltp,
        close: ltp,
        volume: 0,
      },
    ], key);
  }

  const prevClose = daily.length ? daily[daily.length - 1].close : ltp;
  const bar: OHLCVCandle = {
    time: istSessionOpenUnix(key),
    open: prevClose,
    high: Math.max(prevClose, ltp),
    low: Math.min(prevClose, ltp),
    close: ltp,
    volume: 0,
  };
  return [...daily, bar];
}

/** @deprecated Use mergeIntradayForDay with todayIstDayKey() */
export function mergeLiveTodayIntoDaily(
  daily: OHLCVCandle[],
  todayIntraday: OHLCVCandle[],
): OHLCVCandle[] {
  return mergeIntradayForDay(daily, todayIntraday, todayIstDayKey());
}

export function enrichDailyWithLiveToday(
  daily: OHLCVCandle[],
  intraday: OHLCVCandle[],
  ltp?: number | null,
): OHLCVCandle[] {
  const todayKey = todayIstDayKey();
  let merged = mergeIntradayForDay(daily, intraday, todayKey);
  if (!dailySeriesMissingToday(merged)) return merged;
  if (ltp != null && ltp > 0) {
    merged = appendTodayFromLtp(merged, ltp, todayKey);
  }
  return merged;
}

export function parseLtpFromUpstoxJson(data: unknown, instrumentKey: string): number | null {
  if (!data || typeof data !== "object") return null;
  const root = data as Record<string, unknown>;
  const payload = (root.data ?? root) as Record<string, unknown>;
  if (!payload || typeof payload !== "object") return null;

  let quote = payload[instrumentKey] as Record<string, unknown> | undefined;
  if (!quote) {
    const values = Object.values(payload);
    quote = values.find((v) => v && typeof v === "object") as Record<string, unknown> | undefined;
  }
  if (!quote) return null;

  for (const key of ["last_price", "lastPrice", "ltp", "close"]) {
    const val = quote[key];
    if (val != null && Number.isFinite(Number(val))) return Number(val);
  }
  return null;
}

File: proxy_server/chart_live_daily.py
"""Append/replace today's daily bar from intraday session or Upstox LTP."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from zoneinfo import ZoneInfo

import aiohttp

from . import upstox_sdk
from .config import resolve_instrument_key

IST = ZoneInfo("Asia/Kolkata")


def ist_day_key(unix_sec: int) -> str:
    return datetime.fromtimestamp(unix_sec, tz=IST).strftime("%Y-%m-%d")


def today_ist_key() -> str:
    return datetime.now(IST).strftime("%Y-%m-%d")


def ist_session_open_unix(day_key: str) -> int:
    y, m, d = (int(x) for x in day_key.split("-"))
    dt = datetime(y, m, d, 9, 15, tzinfo=IST)
    return int(dt.timestamp())


def daily_missing_today(daily: list[dict[str, Any]]) -> bool:
    if not daily:
        return True
    return ist_day_key(int(daily[-1]["time"])) != today_ist_key()


def merge_intraday_session_for_day(
    daily: list[dict[str, Any]],
    intraday: list[dict[str, Any]],
    day_key: str,
) -> list[dict[str, Any]]:
    if not intraday:
        return daily

    session = [c for c in intraday if ist_day_key(int(c["time"])) == day_key]
    if not session:
        return daily

    session.sort(key=lambda x: int(x["time"]))
    o = session[0]["open"]
    h = max(float(c["high"]) for c in session)
    l = min(float(c["low"]) for c in session)
    cl = session[-1]["close"]
    vol = sum(float(c.get("volume") or 0) for c in session)

    existing = next((c for c in daily if ist_day_key(int(c["time"])) == day_key), None)
    bar_time = int(existing["time"]) if existing else ist_session_open_unix(day_key)
    bar = {
        "time": bar_time,
        "open": float(o),
        "high": h,
        "low": l,
        "close": cl,
        "volume": vol,
    }
    history = [c for c in daily if ist_day_key(int(c["time"])) != day_key]
    merged = history + [bar]
    merged.sort(key=lambda c: int(c["time"]))
    return merged


def append_today_from_ltp(
    daily: list[dict[str, Any]],
    ltp: float,
    day_key: str | None = None,
) -> list[dict[str, Any]]:
    key = day_key or today_ist_key()
    if any(ist_day_key(int(c["time"])) == key for c in daily):
        return merge_intraday_session_for_day(
            daily,
            [
                {
                    "time": ist_session_open_unix(key),
                    "open": ltp,
                    "high": ltp,
                    "low": ltp,
                    "close": ltp,
                    "volume": 0,
                }
            ],
            key,
        )

    prev_close = float(daily[-1]["close"]) if daily else ltp
    o = prev_close
    bar = {
        "time": ist_session_open_unix(key),
        "open": o,
        "high": max(o, ltp),
        "low": min(o, ltp),
        "close": ltp,
        "volume": 0,
    }
    return [*daily, bar]


def parse_ltp_from_upstox_response(data: Any, instrument_key: str) -> float | None:
    if not isinstance(data, dict):
        return None
    payload = data.get("data", data)
    if not isinstance(payload, dict):
        return None
    quote = payload.get(instrument_key) or payload.get(instrument_key.upper())
    if not isinstance(quote, dict) and payload:
        quote = next(iter(payload.values()), None)
    if not isinstance(quote, dict):
        return None
    for key in ("last_price", "lastPrice", "ltp", "close"):
        val = quote.get(key)
        if val is not None:
            try:
                return float(val)
            except (TypeError, ValueError):
                continue
    return None


async def fetch_symbol_ltp(
    session: aiohttp.ClientSession,
    symbol: str,
    user_access_token: str | None,
) -> float | None:
    instrument_key = resolve_instrument_key(symbol.upper())
    if not instrument_key:
        return None
    try:
        raw = await upstox_sdk.get_ltp(user_access_token, instrument_key)
        return parse_ltp_from_upstox_response(raw, instrument_key)
    except RuntimeError:
        return None


async def enrich_daily_candles(
    daily: list[dict[str, Any]],
    intraday: list[dict[str, Any]],
    *,
    session: aiohttp.ClientSession | None = None,
    symbol: str | None = None,
    user_access_token: str | None = None,
) -> list[dict[str, Any]]:
    """Ensure the series includes today's IST bar (intraday aggregate, else LTP)."""
    today_key = today_ist_key()
    merged = merge_intraday_session_for_day(daily, intraday, today_key)
    if not daily_missing_today(merged):
        return merged
    if session and symbol:
        ltp = await fetch_symbol_ltp(session, symbol, user_access_token)
        if ltp is not None and ltp > 0:
            return append_today_from_ltp(merged, ltp, today_key)
    return merged
